### PR TITLE
Experimental features: run with external-environment and project archive download/extract

### PR DESCRIPTION
**Pull request by AlbertDeFusco**
Two experimental features are added here

## External environment 
You can run commands using conda environments not managed by the project itself.

`conda project run --external-environment=<env> [command] [args ...]`

Where `<env>` can be a relative or absolute path to a conda environment, or the name of a global environment stored in one of `envs_dirs` directories.

If there is a collision between a local environment directory and named environment, the local environment directory is preferred.

If no local environment directory found it will check through the conda config "envs_dirs" in order and return the first matching named environment.

~In this PR a defined environment is not required.~

## Download and extract a project archive

Two new flags are added `--project-archive` and `--archive-storage-options`

```
--project-archive PROJECT_ARCHIVE_FILE_OR_URL
                        EXPERIMENTAL: Extract and run directly from a project archive. The archive can be a local file or a
                        fsspec compatible URL. You may need to install appropriate driver packages to work with remote
                        archives. Optionally, use --directory to set the destination directory of the extracted project.
--archive-storage-options ARCHIVE_STORAGE_OPTIONS
                        EXPERIMENTAL: Comma separated list of fsspec storage_options for accessing a remote archive For example
                        --archive-storage-options username=<user>,password=<pass>
```

These flags are enabled for
* lock
* check
* install
* activate
* run

TODO:

- [x] error handling and messages
- [x] tests
- [ ] docs (maybe)

### CODE REVIEW

**Comment by AlbertDeFusco** [comment]
Added support for the currently activated env if a) the project has no defined environment and b) `--external-environment` is not supplied

**Comment by AlbertDeFusco** [comment]
Added support for the currently activated env if a) the project has no defined environment and b) `--external-environment` is not supplied

File: pyproject.toml [hunk 1]
```diff
@@ -38,6 +38,8 @@ dependencies = [
     "pydantic",
     "shellingham",
     "python-dotenv",
+    "fsspec",
+    "python-libarchive-c"
```
Comment (philippjfr) [inline pyproject.toml:42]: We were just testing this and on pypi the package is called:

```suggestion
    "libarchive-c"
```

**Comment by AlbertDeFusco** [comment]
Updates:

* `--project-archive <local-or-remote-archive` supports local files or fsspec-compatible urls to automatically download and extract projects
    * the archive can be any format and compression supported by [libarchive](https://github.com/libarchive/libarchive/wiki/LibarchiveFormats)
* `--directory` combined with `--project-archive` will control the output project directory upon extraction, without the `--directory` parameter the project directly will be
    * the top-level directory in the archive
    * if no top-level directory is present the basename of the archive becomes the name of the project directory
* added `--archive-storage-options <key>=<value>,<key>,<value>...` is a comma separated list of fsspec storage_options parameters
* `--project-archive` works for check, install, lock, activate, and run
    * the project archive is extracted (see `--directory`) and the command is executed against the extracted project directory


for example

```
conda install defusco/label/dev::conda-project=0.2.0.dev aiohttp requests
```

```
conda-project run --project-archive https://anaconda.org/defusco/project/windturbines/download jupyter notebook wind-turbines.ipynb
```

File: README.md [hunk 2]
```diff
@@ -17,6 +16,12 @@ discussion of the differences between "It works for me." to "I've made sure that
 Conda Project is a framework that aims to help you to ensure a high degree of reproducibility in the projects you
 create.
 
+### How is this different from Anaconda Project?
+
+This package is intended as a successor to [Anaconda Project](https://github.com/Anaconda-Platform/anaconda-project).
+We chose to create Conda Project to foster community involvement, adopt newer standards like conda-lock, and provide
+a conda-native workflow. A [standalone conversion script](https://gist.github.com/AlbertDeFusco/e1bc5b960340e78ac9b563fc8de5c727) is provided.
```
Comment (mattkram) [inline README.md:23]: Should we instead include this in a `scripts/` directory in this repo?

**Review by mattkram — APPROVED** [review]
A few minor comments, but LGTM. I fixed a few typos in the docs.

File: src/conda_project/project.py [hunk 3]
```diff
@@ -130,6 +139,69 @@ def __init__(self, directory: Union[Path, str] = "."):
...
+        fn: Union[Path, str],
+        storage_options: Optional[Dict[str, str]] = None,
+        output_directory: Union[Path, str] = ".",
+    ):
+        """Extra a conda-project archive and load the project"""
+
+        if isinstance(output_directory, str):
+            output_directory = Path(output_directory)
+
+        storage_options = {} if storage_options is None else storage_options
```
[the remark this answers is not in the document]
Comment (AlbertDeFusco) [reply src/conda_project/project.py:154]: I'm not sure I follow. The `from_archive` method declares

```python
storage_options: Optional[Dict[str, str]] = None
```

I suppose I'm protecting against use of this function outside of the cli. Does that make sense?

[the remark this answers is not in the document]
Comment (mattkram) [reply src/conda_project/project.py:154]: Yeah, I think it's fine. Just recognized that maybe you could pass in the `None` from the CLI and simplify that code slightly, but it's really a nit.

[the remark this answers is not in the document]
Comment (AlbertDeFusco) [reply src/conda_project/project.py:154]: Yeah, the `key1=valueA,key1=valueB` to `{'key1': 'valueA', 'key2': 'valueB'}` conversion step takes more code than I might like. Maybe there is a better way. I'll give it some more thought. Thanks

**Pull request merged**